### DISCUSSION TITLE
Use `jpackage` to generate native apps

### DISCUSSION THREAD

**Discussion by fvarrui** [fvarrui/JavaPackager#292] [Ideas]
Since I haven't used the `jpackage` tool much, I'd like to know if using `jpackage` to generate native Java apps (instead of different plugins and libraries) would be a good idea.

**Comment by SaptarshiSarkar12** [comment]
@fvarrui Yes, jpackage is far better than jlink as it generates native installers which manages the JRE and dependencies part, so there is no overhead on the end user.
References :
1. https://docs.oracle.com/en/java/javase/20/docs/specs/man/jpackage.html
2. https://www.devdungeon.com/content/use-jpackage-create-native-java-app-installers
3. https://www.baeldung.com/java14-jpackage
4. **https://www.dariawan.com/tutorials/java/java-14-packaging-tool-jep-343/**

**Reply by SaptarshiSarkar12** [reply]
You can also use jpackage to create installers for CLI applications. See here - https://www.dariawan.com/tutorials/java/java-14-packaging-tool-jep-343/

**Reply by fvarrui** [reply]
Ok, thanks for your comments! AFAIK, `jpackage` is only available since JDK 14 ... maybe some artifact generators of JavaPackager could be remaked using `jpackage` when JDK is 14+.

**Reply by fvarrui** [reply]
There's already a jpackage plugin for Maven: https://github.com/akman/jpackage-maven-plugin

**Reply by SaptarshiSarkar12** [reply]
Ok, then it can be included as a plugin in JavaPackager pom file. Will it be okay @fvarrui?

**Reply by fvarrui** [reply]
We'd need another one for Gradle... anyway, it seems it's not hard to run `jpackage` from the command line. Maybe `jpackage-maven-plugin` will better meet your needs?

**Reply by SaptarshiSarkar12** [reply]
@fvarrui Yeah, it can be. But, jpackage builds CLI installers only for windows and not for all Operating Systems.
What about trying GraalVM? Can it solve the problem?
> I don't have any experience about GraalVM

**Reply by fvarrui** [reply]
No idea

> I don't have any experience about GraalVM

Neither do I

**Reply by SaptarshiSarkar12** [reply]
@fvarrui I have tried GraalVM after I started learning it. It is just awesome. It builds pure native executables for all OS, with almost zero dependency requirements. It can also build installers (For JavaFX app) for all OS (deb and rpm for Linux is under development).
You can check [my pom file here](https://github.com/SaptarshiSarkar12/Drifty/blob/eb9fedd0420651481735fd508d04a1cab886fb85/pom.xml) and the [GitHub Actions to build the executables here](https://github.com/SaptarshiSarkar12/Drifty/actions/runs/6079926278). You can download the artifacts from there and test it in your machine.
## Screenshots of Installer on macOS
<p float="left">
<img width="49%" alt="1" src="https://github.com/fvarrui/JavaPackager/assets/105960032/23f69f5a-836c-4cd3-8e05-5ad4ed1a0b6d">
<img width="49%" alt="2" src="https://github.com/fvarrui/JavaPackager/assets/105960032/94a14bfe-8ffd-42ac-a615-3029a0c16150">
</p>
<p float="left">
<img width="49%" alt="3" src="https://github.com/fvarrui/JavaPackager/assets/105960032/2461a701-4b56-41b1-bbd3-35fd70921bd8">
<img width="49%" alt="4" src="https://github.com/fvarrui/JavaPackager/assets/105960032/acb5a1e6-22e2-4391-bc71-5492c548b8ce">
</p>

**Reply by SaptarshiSarkar12** [reply]
@fvarrui I am sorry to say that I have moved to GraalVM from JavaPackager because of its native image generation capabilities. So, I would request you to remove the [Drifty](https://github.com/SaptarshiSarkar12/Drifty/) project from the list of projects which use JavaPackager.

**Reply by fvarrui** [reply]
Hi @SaptarshiSarkar12! Don't worry. Thanks so much for let me know and happy to know that you found a solution that meets your needs. There are more devs asking for JavaPackager supports GraalVM. I haven't tried yet, but I'll probably have to do it. Maybe GraalVM turns JP into a deprecated project

**Reply by SaptarshiSarkar12** [reply]
Thank you for understanding! Yes, GraalVM is indeed going to turn JP into a deprecated project. It will help other people using 
JavaPackager if you add support for GraalVM.
@fvarrui I would also like to inform you that for building native executables for any JavaFX app, _there is a maven plugin named [**GluonFX Maven Plugin**](https://github.com/gluonhq/gluonfx-maven-plugin/) which can generate **JavaFX native executables** using GraalVM under the hood_.
